fetch-in-react-practice/client/components/app.jsx

```diff
@@ -0,0 +1,114 @@
+      .then(response => response.json())
+      .then(todos => this.setState({ todos }));
+  }
+
+  addTodo(newTodo) {
+    /**
+    * Use fetch to send a POST request to `/api/todos`.
+    * Then 😉, once the response JSON is received and parsed,
+    * add the created todo to the state array.
+    *
+    * Do not mutate the original state array, nor any objects within it.
+    * https://reactjs.org/docs/optimizing-performance.html#the-power-of-not-mutating-data
+    *
+    * TIP: Be sure to SERIALIZE the todo object in the body with JSON.stringify()
+    * and specify the "Content-Type" header as "application/json"
+    *
+    * TIP: Use Array.prototype.concat to create a new array containing the contents
+    * of the old array, plus the object returned by the server.
+    */
+    fetch('/api/todos', {
+      method: 'POST',
+      headers: { 'Content-Type': 'application/json' },
+      body: JSON.stringify(newTodo)
+    })
+      .then(response => response.json())
+      .then(todos => this.setState(state => ({ todos: state.todos.concat(todos) })));
+  }
+
+  toggleCompleted(todoId) {
+    /**
+     * Find the index of the todo with the matching todoId in the state array.
+     * Get its "isCompleted" status.
+     * Make a new object containing the opposite "isCompleted" status.
+     * Use fetch to send a PATCH request to `/api/todos/${todoId}`
+     * Then 😉, once the response JSON is received and parsed,
+     * replace the old todo in the state array.
+     *
+     * NOTE: "toggle" means to flip back and forth, so clicking a todo
+     * in the list should "toggle" its isCompleted status back and forth.
+     *
+     * Do not mutate the original state array, nor any objects within it.
+     * https://reactjs.org/docs/optimizing-performance.html#the-power-of-not-mutating-data
+     *
+     * TIP: Be sure to SERIALIZE the updates in the body with JSON.stringify()
+     * And specify the "Content-Type" header as "application/json"
+     */
+    const arr = this.state.todos;
+    const index = this.state.todos.findIndex(x => x.todoId === todoId);
+    let obj = this.state.todos.find(x => x.todoId === todoId);
+    if (obj.isCompleted === false) {
+      const now = { isCompleted: true };
+      fetch(`api/todos/${todoId}`, {
+        method: 'PATCH',
+        headers: { 'Content-Type': 'application/json' },
+        body: JSON.stringify(now)
+      })
+        .then(res => res.json())
+        .then(todos => {
+          obj = todos;
+          arr[index] = obj;
+          this.setState({ todos: arr });
+        });
+    } else if (obj.isCompleted === true) {
+      const now = { isCompleted: false };
+      fetch(`api/todos/${todoId}`, {
+        method: 'PATCH',
+        headers: { 'Content-Type': 'application/json' },
+        body: JSON.stringify(now)
+      })
+        .then(res => res.json())
+        .then(todos => {
+          obj = todos;
+          arr[index] = obj;
+          this.setState({ todos: arr });
+        });
+    }
```

Comment:
Instead of duplicating effort here, practically the exact same thing within the if and else, do this to create your `now` object and get rid of the if statement:

```JAVASCRIPT
const now = { isCompleted: !obj.isCompleted };
```